{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "515 duplicate sequences removed.\n"
     ]
    }
   ],
   "source": [
    "def fasta_parse(fasta_file: str, comment='#'):\n",
    "    \"\"\"\n",
    "    Parses a FASTA file and extracts sequence names and their corresponding sequences.\n",
    "\n",
    "    The function reads a given FASTA file, where each sequence is identified by a line starting with \n",
    "    the '>' symbol, followed by the sequence name. The subsequent lines contain the sequence data, \n",
    "    which is concatenated into a single string for each sequence. The function also ignores any comment \n",
    "    lines that start with the specified comment character.\n",
    "\n",
    "    Parameters:\n",
    "\n",
    "    fasta_file : str\n",
    "        The path to the FASTA file to be parsed.\n",
    "    comment : str, optional\n",
    "        A character indicating comment lines that should be ignored. Default is '#'.\n",
    "    \n",
    "    Returns:\n",
    "    -------\n",
    "    tuple:\n",
    "        A tuple containing two lists:\n",
    "        - names : List[str]\n",
    "            A list of sequence names extracted from lines starting with '>'.\n",
    "        - sequences : List[str]\n",
    "            A list of corresponding sequences, with each sequence represented as a string.\n",
    "    \n",
    "    Example:\n",
    "    --------\n",
    "    >>> names, sequences = fasta_parse(\"example.fasta\")\n",
    "    >>> print(names)\n",
    "    ['sequence1', 'sequence2']\n",
    "    >>> print(sequences)\n",
    "    ['ATCGATCG', 'GGCTAAGT']\n",
    "\n",
    "    Notes:\n",
    "    ------\n",
    "    - Sequence names are taken from lines starting with '>', with the '>' character removed.\n",
    "    - Sequences are converted to uppercase.\n",
    "    - The function assumes that the sequences are stored in a standard FASTA format.\n",
    "\n",
    "    \"\"\"\n",
    "    names = []\n",
    "    sequences = []\n",
    "    name = None\n",
    "    sequence = []\n",
    "    with open(fasta_file, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith(comment):\n",
    "                continue\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                if name is not None:\n",
    "                    names.append(name)\n",
    "                    sequences.append(''.join(sequence))\n",
    "                name = line[1:]\n",
    "                sequence = []\n",
    "            else:\n",
    "                sequence.append(line.upper())\n",
    "        if name is not None:\n",
    "            names.append(name)\n",
    "            sequences.append(''.join(sequence))\n",
    "\n",
    "    return names, sequences\n",
    "\n",
    "def dedup_sequences(data_tuple):\n",
    "    \"\"\"\n",
    "    Removes duplicate sequences from a tuple containing sequence names and sequences.\n",
    "\n",
    "    This function takes a tuple consisting of two lists: one for sequence names and \n",
    "    one for the corresponding sequences. It removes duplicate sequences and returns \n",
    "    a new tuple containing only the unique sequences and their corresponding names.\n",
    "\n",
    "    Parameters:\n",
    "    ----------\n",
    "    data_tuple : tuple\n",
    "        A tuple containing two lists:\n",
    "        - names (list of str): A list of sequence names.\n",
    "        - sequences (list of str): A list of nucleotide or protein sequences.\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    tuple\n",
    "        A tuple containing two lists:\n",
    "        - result_names (list of str): A list of names corresponding to the unique sequences.\n",
    "        - result_sequences (list of str): A list of unique sequences.\n",
    "\n",
    "    Example:\n",
    "    --------\n",
    "    >>> names = [\"seq1\", \"seq2\", \"seq3\"]\n",
    "    >>> sequences = [\"AGCT\", \"CGTA\", \"AGCT\"]\n",
    "    >>> data_tuple = (names, sequences)\n",
    "    >>> dedup_sequences(data_tuple)\n",
    "    ([\"seq1\", \"seq2\"], [\"AGCT\", \"CGTA\"])\n",
    "\n",
    "    This function can also be modified to produce a tuple of duplicate values.\n",
    "    \"\"\"\n",
    "    names, sequences = data_tuple\n",
    "    unique_sequences = {}\n",
    "    result_names = []\n",
    "    result_sequences = []\n",
    "    duplicate_names = []\n",
    "    duplicate_sequences = []\n",
    "    \n",
    "    for name, seq in zip(names, sequences):\n",
    "        if seq not in unique_sequences:\n",
    "            unique_sequences[seq] = name\n",
    "            result_names.append(name)\n",
    "            result_sequences.append(seq)\n",
    "        else:\n",
    "            # these values are actually not used and return at this stage\n",
    "            duplicate_names.append(name)\n",
    "            duplicate_sequences.append(seq)\n",
    "\n",
    "    print(f\"{len(duplicate_sequences)} duplicate sequences removed.\")\n",
    "    return (result_names, result_sequences)\n",
    "\n",
    "def write_fasta(data_tuple, output_fasta):\n",
    "    \"\"\"\n",
    "    Writes a tuple of names and sequences to a FASTA file.\n",
    "    \n",
    "    Parameters:\n",
    "    ----------\n",
    "    data_tuple : tuple\n",
    "        A tuple where the first element is a list of names and the second element is a list of sequences.\n",
    "    output_fasta : str\n",
    "        The name of the output FASTA file.\n",
    "    \"\"\"\n",
    "    names, sequences = data_tuple\n",
    "    with open(output_fasta, 'w') as fasta_file:\n",
    "        for name, sequence in zip(names, sequences):\n",
    "            fasta_file.write(f\">{name}\\n\")\n",
    "            fasta_file.write(f\"{sequence}\\n\")\n",
    "\n",
    "names, sequences = fasta_parse(\"data.fasta\")\n",
    "data_tuple = dedup_sequences((names, sequences))\n",
    "\n",
    "# for archival, hopefully we don't have to read from here again\n",
    "write_fasta(data_tuple, \"dedup_data.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Create directory clustering_tmp\n",
      "easy-cluster --min-seq-id 0.9 -c 0.8 --threads 8 dedup_data.fasta data clustering_tmp \n",
      "\n",
      "MMseqs Version:                     \t15-6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t8\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.9\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb dedup_data.fasta clustering_tmp/6228003922716099929/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[3435] 0s 2ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Nucleotide\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "Create directory clustering_tmp/6228003922716099929/clu_tmp\n",
      "cluster clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu clustering_tmp/6228003922716099929/clu_tmp --max-seqs 20 -c 0.8 --spaced-kmer-mode 1 --threads 8 --alignment-mode 3 -e 0.001 --min-seq-id 0.9 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 1\n",
      "linclust clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/clu_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 10000 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:21,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:21,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 10000 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:21,nucl:5 --min-seq-id 0.9 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 10000 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 3460 type: Nucleotide\n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 3.46K 0s 28ms     \n",
      "\n",
      "Adjusted k-mer length 17\n",
      "Sort kmer 0h 0m 0s 5ms\n",
      "Sort by rep. sequence 0h 0m 0s 2ms\n",
      "Time for fill: 0h 0m 0s 1ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 49ms\n",
      "rescorediagonal clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 3.46K 0s 9ms                                          ] 14.22% 493 eta 0s       \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "clust clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref_rescore1 clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pre_clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 3.46K 0s 5ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 2786 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 3.46K 0s 5ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 3.46K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 13ms\n",
      "Total time: 0h 0m 0s 14ms\n",
      "\n",
      "Size of the sequence database: 3460\n",
      "Size of the alignment database: 3460\n",
      "Number of clusters: 2151\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "createsubdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/order_redundancy clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "createsubdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/order_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "filterdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref_filter1 clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref_filter2 --filter-file clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/order_redundancy --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 2.15K 0s 8ms      \n",
      "Time for merging to pref_filter2: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "align clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/input_step_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/input_step_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref_filter2 clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 10000 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 2151 type: Nucleotide\n",
      "Target database size: 2151 type: Nucleotide\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 2.15K 0s 18ms     \n",
      "Time for merging to aln: 0h 0m 0s 1ms\n",
      "5781 alignments calculated\n",
      "2426 sequence pairs passed the thresholds (0.419651 of overall calculated)\n",
      "1.127848 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 35ms\n",
      "clust clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/input_step_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/aln clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 2.15K 0s 4ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 275 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 2.15K 0s 5ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 2.15K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 11ms\n",
      "Total time: 0h 0m 0s 12ms\n",
      "\n",
      "Size of the sequence database: 2151\n",
      "Size of the alignment database: 2151\n",
      "Number of clusters: 1935\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 13ms\n",
      "mergeclusters clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/clu_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pre_clust clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/clust --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 2.15K 0s 4ms      \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 1.93K 0s 8ms      \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 3.46K 0s 12ms     \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/linclust/16156693663226617389/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/clu_redundancy clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "extractframes clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/input_step_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/query_seqs --forward-frames 1 --reverse-frames 1 --create-lookup 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 1.93K 0s 6ms      \n",
      "Time for merging to query_seqs_h: 0h 0m 0s 1ms\n",
      "Time for merging to query_seqs: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 15ms\n",
      "prefilter clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/query_seqs clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/input_step_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 15 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 10000 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 1 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 60 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 3870 type: Nucleotide\n",
      "Estimated memory consumption: 8G\n",
      "Target database size: 1935 type: Nucleotide\n",
      "Index table k-mer threshold: 0 at k-mer size 15 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 1.93K 0s 70ms     \n",
      "Index table: Masked residues: 882\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 1.93K 0s 14ms     \n",
      "Index statistics\n",
      "Entries:          435970\n",
      "DB size:          8194 MB\n",
      "Avg k-mer size:   0.000406\n",
      "Top 10 k-mers\n",
      "    TGAAGGTGGGGGCCA\t108\n",
      "    AAGGGAAAGGGTAGA\t105\n",
      "    GGAATCCGGCAAGGC\t75\n",
      "    CCACGGAGAACAAAG\t54\n",
      "    AAGCGGGCCCGGCGG\t45\n",
      "    GGGGAAAGTGTAGGC\t35\n",
      "    GGGGAAGGTGTAGGC\t34\n",
      "    AGGTAAAGGTGAAAG\t34\n",
      "    AAATCGGGTCAGGAG\t33\n",
      "    AGTCGGCTCAGCAGG\t32\n",
      "Time for index table init: 0h 0m 3s 258ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 0\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 3870\n",
      "Target db start 1 to 1935\n",
      "[=================================================================] 100.00% 3.87K 0s 26ms     \n",
      "\n",
      "0.846582 k-mers per position\n",
      "234 DB matches per sequence\n",
      "0 overflows\n",
      "0 sequences passed prefiltering per query sequence\n",
      "0 median result list length\n",
      "2301 sequences with 0 size result lists\n",
      "Time for merging to pref: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 3s 401ms\n",
      "rescorediagonal clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/query_seqs clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/input_step_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/pref clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln_ungapped --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 2 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 3.87K 0s 6ms      \n",
      "Time for merging to aln_ungapped: 0h 0m 0s 1ms===================>] 98.60% 3.82K eta 0s       \n",
      "Time for processing: 0h 0m 0s 12ms\n",
      "subtractdbs clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/pref clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln_ungapped clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/pref_subtract --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "subtractdbs clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/pref clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln_ungapped clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/pref_subtract --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Remove clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln_ungapped ids from clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/pref\n",
      "[=================================================================] 100.00% 3.87K 0s 5ms      \n",
      "Time for merging to pref_subtract: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "align clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/query_seqs clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/input_step_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/pref_subtract clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln_gapped --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.9 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 10000 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 3870 type: Nucleotide\n",
      "Target database size: 1935 type: Nucleotide\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 3.87K 0s 5ms      \n",
      "Time for merging to aln_gapped: 0h 0m 0s 1ms\n",
      "412 alignments calculated\n",
      "34 sequence pairs passed the thresholds (0.082524 of overall calculated)\n",
      "0.008786 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 22ms\n",
      "concatdbs clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln_ungapped clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln_gapped clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln --preserve-keys --take-larger-entry --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 3.87K 0s 6ms      2.88K eta 0s       \n",
      "[=================================================================] 100.00% 3.87K 0s 11ms     \n",
      "Time for merging to aln: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 15ms\n",
      "offsetalignment clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/input_step_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/query_seqs clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/input_step_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/input_step_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln_off --chain-alignments 0 --merge-query 1 --search-type 3 --threads 8 --compressed 0 --db-load-mode 0 -v 3 \n",
      "\n",
      "Computing ORF lookup\n",
      "Computing contig offsets\n",
      "Computing contig lookup\n",
      "Time for contig lookup: 0h 0m 0s 0ms\n",
      "Writing results to: clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln_off\n",
      "[=================================================================] 100.00% 3.46K 0s 6ms      \n",
      "\n",
      "Time for merging to aln_off: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "clust clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/input_step_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln_off clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/clu --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 1.93K 0s 4ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 10 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 1.93K 0s 3ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 1.93K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 9ms\n",
      "Total time: 0h 0m 0s 10ms\n",
      "\n",
      "Size of the sequence database: 1935\n",
      "Size of the alignment database: 1935\n",
      "Number of clusters: 1918\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "mergeclusters clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/clu_redundancy clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/clu --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 1.93K 0s 4ms      \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 1.92K 0s 10ms     9 eta 0s       \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 3.46K 0s 14ms     \n",
      "Time for merging to clu: 0h 0m 0s 1ms===========================> ] 97.20% 3.36K eta 0s       \n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/query_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/perf -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/aln_off -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_tmp/10486537641019549264/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu clustering_tmp/6228003922716099929/cluster.tsv --threads 8 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "result2repseq clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu clustering_tmp/6228003922716099929/clu_rep --db-load-mode 0 --compressed 0 --threads 8 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 1.92K 0s 5ms      ========================================>                    ] 68.81% 1.32K eta 0s       \n",
      "Time for merging to clu_rep: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 13ms\n",
      "result2flat clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu_rep clustering_tmp/6228003922716099929/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "createseqfiledb clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu clustering_tmp/6228003922716099929/clu_seqs --threads 8 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 1.92K 0s 5ms      \n",
      "Time for merging to clu_seqs: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "result2flat clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/input clustering_tmp/6228003922716099929/clu_seqs clustering_tmp/6228003922716099929/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering_tmp/6228003922716099929/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/6228003922716099929/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster --min-seq-id 0.9 -c 0.8 --threads 8 dedup_data.fasta data clustering_tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster --min-seq-id 0.85 -c 0.7 --threads 8 dedup_data.fasta data_less_strict clustering_tmp \n",
      "\n",
      "MMseqs Version:                     \t15-6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.7\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t8\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.85\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb dedup_data.fasta clustering_tmp/7737306139982184380/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[3435] 0s 1ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Nucleotide\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "Create directory clustering_tmp/7737306139982184380/clu_tmp\n",
      "cluster clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu clustering_tmp/7737306139982184380/clu_tmp --max-seqs 20 -c 0.7 --spaced-kmer-mode 1 --threads 8 --alignment-mode 3 -e 0.001 --min-seq-id 0.85 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 1\n",
      "linclust clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/clu_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.85 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.7 --cov-mode 0 --max-seq-len 10000 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:21,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:21,nucl:5 --min-seq-id 0.85 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.7 --max-seq-len 10000 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:21,nucl:5 --min-seq-id 0.85 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.7 --max-seq-len 10000 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 3460 type: Nucleotide\n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 3.46K 0s 31ms     \n",
      "\n",
      "Adjusted k-mer length 17\n",
      "Sort kmer 0h 0m 0s 6ms\n",
      "Sort by rep. sequence 0h 0m 0s 2ms\n",
      "Time for fill: 0h 0m 0s 1ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 50ms\n",
      "rescorediagonal clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.7 -a 0 --cov-mode 0 --min-seq-id 0.85 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 3.46K 0s 8ms      \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 1ms==================>] 99.05% 3.43K eta 0s       \n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "clust clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref_rescore1 clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pre_clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 3.46K 0s 6ms                                          ] 20.61% 714 eta 0s       \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 4080 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 3.46K 0s 6ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 3.46K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 14ms\n",
      "Total time: 0h 0m 0s 15ms\n",
      "\n",
      "Size of the sequence database: 3460\n",
      "Size of the alignment database: 3460\n",
      "Number of clusters: 1975\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "createsubdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/order_redundancy clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "createsubdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/order_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "filterdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref_filter1 clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref_filter2 --filter-file clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/order_redundancy --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 1.98K 0s 6ms      \n",
      "Time for merging to pref_filter2: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "align clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/input_step_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/input_step_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref_filter2 clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.85 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.7 --cov-mode 0 --max-seq-len 10000 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 1975 type: Nucleotide\n",
      "Target database size: 1975 type: Nucleotide\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 1.98K 0s 18ms     \n",
      "Time for merging to aln: 0h 0m 0s 1ms\n",
      "5448 alignments calculated\n",
      "2472 sequence pairs passed the thresholds (0.453745 of overall calculated)\n",
      "1.251646 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 36ms\n",
      "clust clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/input_step_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/aln clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 1.98K 0s 4ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 497 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 1.98K 0s 5ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 1.98K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 11ms\n",
      "Total time: 0h 0m 0s 12ms\n",
      "\n",
      "Size of the sequence database: 1975\n",
      "Size of the alignment database: 1975\n",
      "Number of clusters: 1649\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 13ms\n",
      "mergeclusters clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/clu_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pre_clust clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/clust --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 1.98K 0s 4ms      \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 1.65K 0s 9ms      \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 3.46K 0s 13ms     \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/linclust/4815149358541261339/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/clu_redundancy clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "extractframes clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/input_step_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/query_seqs --forward-frames 1 --reverse-frames 1 --create-lookup 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 1.65K 0s 6ms      \n",
      "Time for merging to query_seqs_h: 0h 0m 0s 1ms\n",
      "Time for merging to query_seqs: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "prefilter clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/query_seqs clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/input_step_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 15 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 10000 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.7 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 1 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 60 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 3298 type: Nucleotide\n",
      "Estimated memory consumption: 8G\n",
      "Target database size: 1649 type: Nucleotide\n",
      "Index table k-mer threshold: 0 at k-mer size 15 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 1.65K 0s 32ms     \n",
      "Index table: Masked residues: 818\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 1.65K 0s 11ms     \n",
      "Index statistics\n",
      "Entries:          387313\n",
      "DB size:          8194 MB\n",
      "Avg k-mer size:   0.000361\n",
      "Top 10 k-mers\n",
      "    TGAAGGTGGGGGCCA\t92\n",
      "    AAGGGAAAGGGTAGA\t89\n",
      "    GGAATCCGGCAAGGC\t65\n",
      "    CCACGGAGAACAAAG\t41\n",
      "    AAGCGGGCCCGGCGG\t36\n",
      "    AAGGCACAAAAACGC\t32\n",
      "    GGGGAAAGTGTAGGC\t32\n",
      "    AGGTAAACGTGAAAG\t32\n",
      "    AGGTAAAGGTGAAAG\t28\n",
      "    AAATCGGGTCAGGAG\t28\n",
      "Time for index table init: 0h 0m 3s 1ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 0\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 3298\n",
      "Target db start 1 to 1649\n",
      "[=================================================================] 100.00% 3.30K 0s 23ms     \n",
      "\n",
      "0.849562 k-mers per position\n",
      "216 DB matches per sequence\n",
      "0 overflows\n",
      "0 sequences passed prefiltering per query sequence\n",
      "0 median result list length\n",
      "1971 sequences with 0 size result lists\n",
      "Time for merging to pref: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 3s 146ms\n",
      "rescorediagonal clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/query_seqs clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/input_step_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/pref clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln_ungapped --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 2 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.7 -a 0 --cov-mode 0 --min-seq-id 0.85 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 3.30K 0s 6ms      \n",
      "Time for merging to aln_ungapped: 0h 0m 0s 1ms=============>      ] 89.90% 2.96K eta 0s       \n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "subtractdbs clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/pref clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln_ungapped clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/pref_subtract --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "subtractdbs clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/pref clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln_ungapped clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/pref_subtract --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Remove clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln_ungapped ids from clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/pref\n",
      "[=================================================================] 100.00% 3.30K 0s 5ms      \n",
      "Time for merging to pref_subtract: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "align clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/query_seqs clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/input_step_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/pref_subtract clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln_gapped --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.85 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.7 --cov-mode 0 --max-seq-len 10000 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 3298 type: Nucleotide\n",
      "Target database size: 1649 type: Nucleotide\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 3.30K 0s 4ms      \n",
      "Time for merging to aln_gapped: 0h 0m 0s 1ms\n",
      "228 alignments calculated\n",
      "20 sequence pairs passed the thresholds (0.087719 of overall calculated)\n",
      "0.006064 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 21ms\n",
      "concatdbs clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln_ungapped clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln_gapped clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln --preserve-keys --take-larger-entry --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 3.30K 0s 4ms      \n",
      "[=================================================================] 100.00% 3.30K 0s 7ms      \n",
      "Time for merging to aln: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 12ms\n",
      "offsetalignment clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/input_step_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/query_seqs clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/input_step_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/input_step_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln_off --chain-alignments 0 --merge-query 1 --search-type 3 --threads 8 --compressed 0 --db-load-mode 0 -v 3 \n",
      "\n",
      "Computing ORF lookup\n",
      "Computing contig offsets\n",
      "Computing contig lookup\n",
      "Time for contig lookup: 0h 0m 0s 0ms\n",
      "Writing results to: clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln_off\n",
      "[=================================================================] 100.00% 3.46K 0s 7ms                                   ] 45.27% 1.57K eta 0s       \n",
      "\n",
      "Time for merging to aln_off: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "clust clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/input_step_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln_off clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/clu --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 1.65K 0s 3ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 5 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 1.65K 0s 4ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 1.65K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 10ms\n",
      "Total time: 0h 0m 0s 10ms\n",
      "\n",
      "Size of the sequence database: 1649\n",
      "Size of the alignment database: 1649\n",
      "Number of clusters: 1633\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 12ms\n",
      "mergeclusters clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/clu_redundancy clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/clu --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 1.65K 0s 4ms      \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 1.63K 0s 7ms      \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 3.46K 0s 11ms     \n",
      "Time for merging to clu: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/query_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/perf -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/aln_off -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_tmp/12089824728692127930/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu clustering_tmp/7737306139982184380/cluster.tsv --threads 8 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "result2repseq clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu clustering_tmp/7737306139982184380/clu_rep --db-load-mode 0 --compressed 0 --threads 8 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 1.63K 0s 4ms      \n",
      "Time for merging to clu_rep: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "result2flat clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu_rep clustering_tmp/7737306139982184380/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "createseqfiledb clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu clustering_tmp/7737306139982184380/clu_seqs --threads 8 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 1.63K 0s 5ms      \n",
      "Time for merging to clu_seqs: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 9ms\n",
      "result2flat clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/input clustering_tmp/7737306139982184380/clu_seqs clustering_tmp/7737306139982184380/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "rmdb clustering_tmp/7737306139982184380/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/7737306139982184380/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster --min-seq-id 0.85 -c 0.7 --threads 8 dedup_data.fasta data_less_strict clustering_tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "easy-cluster --min-seq-id 0.95 -c 0.9 --threads 8 dedup_data.fasta data_more_strict clustering_tmp \n",
      "\n",
      "MMseqs Version:                     \t15-6f452\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.9\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t8\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.95\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb dedup_data.fasta clustering_tmp/1128489317349927253/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "\u001b[33mShuffle database cannot be combined with --createdb-mode 0\n",
      "\u001b[39m\u001b[33mWe recompute with --shuffle 0\n",
      "\u001b[39mConverting sequences\n",
      "[3435] 0s 2ms\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "Database type: Nucleotide\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "Create directory clustering_tmp/1128489317349927253/clu_tmp\n",
      "cluster clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu clustering_tmp/1128489317349927253/clu_tmp --max-seqs 20 -c 0.9 --spaced-kmer-mode 1 --threads 8 --alignment-mode 3 -e 0.001 --min-seq-id 0.95 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 1.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 1\n",
      "linclust clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/clu_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.95 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.9 --cov-mode 0 --max-seq-len 10000 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:21,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:21,nucl:5 --min-seq-id 0.95 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.9 --max-seq-len 10000 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:21,nucl:5 --min-seq-id 0.95 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.9 --max-seq-len 10000 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 3460 type: Nucleotide\n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 100.00% 3.46K 0s 19ms     \n",
      "\n",
      "Adjusted k-mer length 17\n",
      "Sort kmer 0h 0m 0s 4ms\n",
      "Sort by rep. sequence 0h 0m 0s 2ms\n",
      "Time for fill: 0h 0m 0s 0ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 34ms\n",
      "rescorediagonal clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.9 -a 0 --cov-mode 0 --min-seq-id 0.95 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 3.46K 0s 9ms      \n",
      "Time for merging to pref_rescore1: 0h 0m 0s 1ms==================>] 99.68% 3.45K eta 0s       \n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "clust clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref_rescore1 clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pre_clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 3.46K 0s 5ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 1498 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 3.46K 0s 5ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 3.46K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 13ms\n",
      "Total time: 0h 0m 0s 14ms\n",
      "\n",
      "Size of the sequence database: 3460\n",
      "Size of the alignment database: 3460\n",
      "Number of clusters: 2515\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "createsubdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/order_redundancy clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "createsubdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/order_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "filterdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref_filter1 clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref_filter2 --filter-file clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/order_redundancy --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 100.00% 2.52K 0s 7ms      \n",
      "Time for merging to pref_filter2: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "align clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/input_step_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/input_step_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref_filter2 clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.95 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.9 --cov-mode 0 --max-seq-len 10000 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 2515 type: Nucleotide\n",
      "Target database size: 2515 type: Nucleotide\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 2.52K 0s 18ms     \n",
      "Time for merging to aln: 0h 0m 0s 1ms\n",
      "6873 alignments calculated\n",
      "2657 sequence pairs passed the thresholds (0.386585 of overall calculated)\n",
      "1.056461 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 36ms\n",
      "clust clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/input_step_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/aln clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 2.52K 0s 4ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 142 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 2.52K 0s 6ms                                    ] 29.59% 745 eta 0s       \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 2.52K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 13ms\n",
      "Total time: 0h 0m 0s 14ms\n",
      "\n",
      "Size of the sequence database: 2515\n",
      "Size of the alignment database: 2515\n",
      "Number of clusters: 2388\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "mergeclusters clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/clu_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pre_clust clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/clust --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 2.52K 0s 4ms      \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 2.39K 0s 9ms      \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 3.46K 0s 14ms     \n",
      "Time for merging to clu_redundancy: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/linclust/4248365688598968765/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/clu_redundancy clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "extractframes clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/input_step_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/query_seqs --forward-frames 1 --reverse-frames 1 --create-lookup 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 2.39K 0s 7ms      \n",
      "Time for merging to query_seqs_h: 0h 0m 0s 2ms\n",
      "Time for merging to query_seqs: 0h 0m 0s 3ms\n",
      "Time for processing: 0h 0m 0s 20ms\n",
      "prefilter clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/query_seqs clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/input_step_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 15 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 10000 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.9 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 1 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 60 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 4776 type: Nucleotide\n",
      "Estimated memory consumption: 8G\n",
      "Target database size: 2388 type: Nucleotide\n",
      "Index table k-mer threshold: 0 at k-mer size 15 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00% 2.39K 0s 45ms     \n",
      "Index table: Masked residues: 969\n",
      "Index table: fill\n",
      "[=================================================================] 100.00% 2.39K 0s 15ms     \n",
      "Index statistics\n",
      "Entries:          509820\n",
      "DB size:          8194 MB\n",
      "Avg k-mer size:   0.000475\n",
      "Top 10 k-mers\n",
      "    TGAAGGTGGGGGCCA\t130\n",
      "    AAGGGAAAGGGTAGA\t123\n",
      "    GGAATCCGGCAAGGC\t93\n",
      "    CATCGAACCGGTAGC\t84\n",
      "    CCACGGAGAACAAAG\t62\n",
      "    AAGCGGGCCCGGCGG\t56\n",
      "    AGGTAAACGTGAAAG\t44\n",
      "    AGGTAAAGGTGAAAG\t44\n",
      "    AGTCGGCTCAGCAGG\t43\n",
      "    AGGAGTCCGGCCAGG\t39\n",
      "Time for index table init: 0h 0m 3s 38ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 0\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 4776\n",
      "Target db start 1 to 2388\n",
      "[=================================================================] 100.00% 4.78K 0s 29ms     \n",
      "\n",
      "0.844678 k-mers per position\n",
      "277 DB matches per sequence\n",
      "0 overflows\n",
      "0 sequences passed prefiltering per query sequence\n",
      "0 median result list length\n",
      "2802 sequences with 0 size result lists\n",
      "Time for merging to pref: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 3s 181ms\n",
      "rescorediagonal clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/query_seqs clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/input_step_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/pref clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln_ungapped --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 2 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.9 -a 0 --cov-mode 0 --min-seq-id 0.95 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 4.78K 0s 9ms      ==========================>         ] 85.24% 4.07K eta 0s       \n",
      "Time for merging to aln_ungapped: 0h 0m 0s 1ms===================>] 99.98% 4.78K eta 0s       \n",
      "Time for processing: 0h 0m 0s 15ms\n",
      "subtractdbs clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/pref clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln_ungapped clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/pref_subtract --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "subtractdbs clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/pref clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln_ungapped clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/pref_subtract --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Remove clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln_ungapped ids from clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/pref\n",
      "[=================================================================] 100.00% 4.78K 0s 7ms      \n",
      "Time for merging to pref_subtract: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "align clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/query_seqs clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/input_step_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/pref_subtract clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln_gapped --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.95 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.9 --cov-mode 0 --max-seq-len 10000 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 4776 type: Nucleotide\n",
      "Target database size: 2388 type: Nucleotide\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00% 4.78K 0s 6ms      \n",
      "Time for merging to aln_gapped: 0h 0m 0s 1ms\n",
      "935 alignments calculated\n",
      "30 sequence pairs passed the thresholds (0.032086 of overall calculated)\n",
      "0.006281 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 23ms\n",
      "concatdbs clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln_ungapped clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln_gapped clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln --preserve-keys --take-larger-entry --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 4.78K 0s 6ms      \n",
      "[=================================================================] 100.00% 4.78K 0s 10ms     \n",
      "Time for merging to aln: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "offsetalignment clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/input_step_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/query_seqs clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/input_step_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/input_step_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln_off --chain-alignments 0 --merge-query 1 --search-type 3 --threads 8 --compressed 0 --db-load-mode 0 -v 3 \n",
      "\n",
      "Computing ORF lookup\n",
      "Computing contig offsets\n",
      "Computing contig lookup\n",
      "Time for contig lookup: 0h 0m 0s 0ms\n",
      "Writing results to: clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln_off\n",
      "[=================================================================] 100.00% 3.46K 0s 7ms      \n",
      "\n",
      "Time for merging to aln_off: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 12ms\n",
      "clust clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/input_step_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln_off clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/clu --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 8 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 100.00% 2.39K 0s 5ms      \n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 11 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 100.00% 2.39K 0s 4ms      \n",
      "Add missing connections\n",
      "[=================================================================] 100.00% 2.39K 0s 1ms      \n",
      "\n",
      "Time for read in: 0h 0m 0s 11ms\n",
      "Total time: 0h 0m 0s 12ms\n",
      "\n",
      "Size of the sequence database: 2388\n",
      "Size of the alignment database: 2388\n",
      "Number of clusters: 2368\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 13ms\n",
      "mergeclusters clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/clu_redundancy clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/clu --threads 8 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 100.00% 2.39K 0s 5ms      \n",
      "Clustering step 2\n",
      "[=================================================================] 100.00% 2.37K 0s 9ms      \n",
      "Write merged clustering\n",
      "[=================================================================] 100.00% 3.46K 0s 12ms     \n",
      "Time for merging to clu: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 15ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/query_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/perf -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/aln_off -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_tmp/6599874562294733578/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu clustering_tmp/1128489317349927253/cluster.tsv --threads 8 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "result2repseq clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu clustering_tmp/1128489317349927253/clu_rep --db-load-mode 0 --compressed 0 --threads 8 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 2.37K 0s 4ms      \n",
      "Time for merging to clu_rep: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "result2flat clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu_rep clustering_tmp/1128489317349927253/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "createseqfiledb clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu clustering_tmp/1128489317349927253/clu_seqs --threads 8 -v 3 \n",
      "\n",
      "[=================================================================] 100.00% 2.37K 0s 5ms      \n",
      "Time for merging to clu_seqs: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "result2flat clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/input clustering_tmp/1128489317349927253/clu_seqs clustering_tmp/1128489317349927253/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "rmdb clustering_tmp/1128489317349927253/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb clustering_tmp/1128489317349927253/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n"
     ]
    }
   ],
   "source": [
    "!mmseqs easy-cluster --min-seq-id 0.95 -c 0.9 --threads 8 dedup_data.fasta data_more_strict clustering_tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster_Rep</th>\n",
       "      <th>Cluster_Member</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tmRNA_Dich.nodo._TRW-35819_1-352</td>\n",
       "      <td>tmRNA_Dich.nodo._TRW-35819_1-352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>srp_Myco.hyop._AE017332</td>\n",
       "      <td>srp_Myco.hyop._AE017332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5s_Fusarium-cerealis-7</td>\n",
       "      <td>5s_Fusarium-cerealis-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5s_Fusarium-cerealis-7</td>\n",
       "      <td>5s_Fusarium-asiaticum-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5s_Fusarium-cerealis-7</td>\n",
       "      <td>5s_Fusarium-lunulosporum-4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Cluster_Rep                    Cluster_Member\n",
       "0  tmRNA_Dich.nodo._TRW-35819_1-352  tmRNA_Dich.nodo._TRW-35819_1-352\n",
       "1           srp_Myco.hyop._AE017332           srp_Myco.hyop._AE017332\n",
       "2            5s_Fusarium-cerealis-7            5s_Fusarium-cerealis-7\n",
       "3            5s_Fusarium-cerealis-7           5s_Fusarium-asiaticum-6\n",
       "4            5s_Fusarium-cerealis-7        5s_Fusarium-lunulosporum-4"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# Load the TSV file into a DataFrame\n",
    "file_path = 'data_cluster.tsv'\n",
    "cluster_df = pd.read_csv(file_path, sep='\\t', header=None, names=['Cluster_Rep', 'Cluster_Member'])\n",
    "\n",
    "# Display the first few rows\n",
    "cluster_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster_Rep</th>\n",
       "      <th>Cluster_Size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>501</th>\n",
       "      <td>5s_Vibrio-vulnificus-1</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>5s_Escherichia-coli-11</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388</th>\n",
       "      <td>5s_Nocardia-farcinica-1</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>5s_Streptomyces-coelicolor-4</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509</th>\n",
       "      <td>5s_Xenopus-laevis-8</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758</th>\n",
       "      <td>RNaseP_SM-A3153</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757</th>\n",
       "      <td>RNaseP_SM-A3048</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>756</th>\n",
       "      <td>RNaseP_SM-A2946</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>755</th>\n",
       "      <td>RNaseP_SM-A2642</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>959</th>\n",
       "      <td>srp_Chlo.tepi._AE006470</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1918 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Cluster_Rep  Cluster_Size\n",
       "501        5s_Vibrio-vulnificus-1            41\n",
       "273        5s_Escherichia-coli-11            35\n",
       "388       5s_Nocardia-farcinica-1            33\n",
       "460  5s_Streptomyces-coelicolor-4            29\n",
       "509           5s_Xenopus-laevis-8            28\n",
       "..                            ...           ...\n",
       "758               RNaseP_SM-A3153             1\n",
       "757               RNaseP_SM-A3048             1\n",
       "756               RNaseP_SM-A2946             1\n",
       "755               RNaseP_SM-A2642             1\n",
       "959       srp_Chlo.tepi._AE006470             1\n",
       "\n",
       "[1918 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by the cluster representative and count the number of members in each cluster\n",
    "cluster_size = cluster_df.groupby('Cluster_Rep').size().reset_index(name='Cluster_Size')\n",
    "\n",
    "# Sort by cluster size to identify the largest clusters\n",
    "cluster_size_sorted = cluster_size.sort_values(by='Cluster_Size', ascending=False)\n",
    "\n",
    "# Display the largest clusters\n",
    "cluster_size_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
